contrib/IECoreAlembic/src/IECoreAlembic/MeshAlgo.cpp

```diff
@@ -32,61 +33,107 @@
 //
 //////////////////////////////////////////////////////////////////////////
 
-#include "IECore/MeshPrimitive.h"
-
-#include "IECoreAlembic/FromAlembicSubDConverter.h"
+#include "IECoreAlembic/ObjectAlgo.h"
+#include "IECoreAlembic/MeshAlgo.h"
+#include "IECoreAlembic/GeomBaseAlgo.h"
 
 using namespace IECore;
 using namespace IECoreAlembic;
 using namespace Alembic::AbcGeom;
 
-FromAlembicSubDConverter::ConverterDescription<FromAlembicSubDConverter> FromAlembicSubDConverter::g_description;
+namespace IECoreAlembic
+{
 
-IE_CORE_DEFINERUNTIMETYPED( FromAlembicSubDConverter );
+namespace MeshAlgo
+{
 
-FromAlembicSubDConverter::FromAlembicSubDConverter( Alembic::Abc::IObject iSubD )
-	:	FromAlembicGeomBaseConverter( "Converts AbcGeom::ISubD objects to IECore::MeshPrimitive objects", iSubD )
+IECore::MeshPrimitivePtr convert( const Alembic::AbcGeom::IPolyMesh &mesh, const Alembic::Abc::ISampleSelector &sampleSelector )
 {
+	const IPolyMeshSchema &iPolyMeshSchema = mesh.getSchema();
+
+	IPolyMeshSchema::Sample sample = iPolyMeshSchema.getValue( sampleSelector );
+
+	IntVectorDataPtr verticesPerFace = new IntVectorData();
+	verticesPerFace->writable().insert(
+		verticesPerFace->writable().begin(),
+		sample.getFaceCounts()->get(),
+		sample.getFaceCounts()->get() + sample.getFaceCounts()->size()
+	);
+
+	IntVectorDataPtr vertexIds = new IntVectorData();
+	vertexIds->writable().insert(
+		vertexIds->writable().begin(),
+		sample.getFaceIndices()->get(),
+		sample.getFaceIndices()->get() + sample.getFaceIndices()->size()
+	);
+
+	V3fVectorDataPtr points = new V3fVectorData();
+	points->writable().resize( sample.getPositions()->size() );
+	memcpy( &(points->writable()[0]), sample.getPositions()->get(), sample.getPositions()->size() * sizeof( Imath::V3f ) );
+
+	MeshPrimitivePtr result = new IECore::MeshPrimitive( verticesPerFace, vertexIds, "linear", points );
+
+	IN3fGeomParam normals = iPolyMeshSchema.getNormalsParam();
+	if( normals.valid() )
+	{
+		GeomBaseAlgo::convertGeomParam( normals, sampleSelector, result.get() );
+	}
+
+	Alembic::AbcGeom::IV2fGeomParam uvs = iPolyMeshSchema.getUVsParam();
+	GeomBaseAlgo::convertUVs( uvs, sampleSelector, result.get() );
+
+	ICompoundProperty arbGeomParams = iPolyMeshSchema.getArbGeomParams();
+	GeomBaseAlgo::convertArbGeomParams( arbGeomParams, sampleSelector, result.get() );
+
+	return result;
 }
 
-IECore::ObjectPtr FromAlembicSubDConverter::doAlembicConversion( const Alembic::Abc::IObject &iObject, const Alembic::Abc::ISampleSelector &sampleSelector, const IECore::CompoundObject *operands ) const
+IECore::MeshPrimitivePtr convert( const Alembic::AbcGeom::ISubD &mesh, const Alembic::Abc::ISampleSelector &sampleSelector )
 {
-	ISubD iSubD( iObject, kWrapExisting );
+	ISubD iSubD( mesh, kWrapExisting );
 	ISubDSchema &iSubDSchema = iSubD.getSchema();
-	
+
```

Comment:
> Is it worth adding an anonymous template function for the common topology, position, uv & AgbGeomParams convert?

Good idea - done in f28b528146397025d92c18d15fd3cd103a8d16e4.

> If there is a single template function then can this be used to bind to the ObjectAlgo::ConverterDescription?

I'd like to keep the two separate `convert( const ISubd & )` and `convert( const IPolyMesh & )` methods as the main public interface in MeshAlgo.h. I find the Alembic API pretty baffling, and having nice concrete signatures that make it clear what type to pass is helpful. I think having ConverterDescription::Converter be a plain function pointer is justified as it encourages this design.